### PR TITLE
Base for random module & testing + fixed some compilation errors & C++11 compilation

### PR DESCRIPTION
**Pull request by nicpalard**

### CODE REVIEW

**Comment by nicpalard** [comment]
I found that in the C++ world #define is not well a good pratice as it's a macro and it's not type safe and it has no scope. In C we use #define because it's the only way to have global variables but I would suggest to use **const** in the future.

```c++
// Not ok
#define MAX_RADIUS 500
// ok
const int MAX_RADIUS = 500;
```

**Comment by mthl** [comment]
The usage of `const int` seems indeed more idiomatic to C++.  What about using it directly in these commits (or adding a TODO if this is not trivial)?

File: enki/RandomWorld.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
+/*!
+   \file RandomWorld.cpp
+   \brief Random Modul for testing
+   \author PALARD Nicolas
+   \date 04/02/2017
+*/
+
```
Comment (mthl) [inline enki/RandomWorld.cpp]: Can you add Copyright notice for you (and maybe for @Krazyxx  if he was involved) + copying statement (like what is done in `tests/enkiTestWorld.cpp`)?
Please move the file doxygen documentation after the includes like what is done in `enki/PhysicalEngine.h`

File: enki/RandomWorld.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    int posx = randomNumber(0, world->w);
+    int posy = randomNumber(0, world->h);
+    o->pos = Point(posx, posy);
+    return o;
+}
+
+Robot* randomRobot(World* world)
+{
+    int type;
+    if(ONLY_THYMIOS)
```
Comment (mthl) [inline enki/RandomWorld.cpp]: nitpick:
```c++
if (ONLY_THYMIOS)
```
a space should be present after the `if`

File: enki/RandomWorld.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/*!
+   \file RandomWorld.h
+   \brief Header of random module.
+   \author PALARD Nicolas
+   \date 04/02/2017
+*/
```
Comment (mthl) [inline enki/RandomWorld.h]: copyright + copying statement + move after includes like previously explain.

File: enki/RandomWorld.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    int thymiosNumber = randomNumber(1, MAX_ROBOTS);
+    for(int i = 0 ; i < thymiosNumber ; i++)
+    {
+        Robot* r = randomRobot(world);
+        world->addObject(r);
+    }
+    // Random number of objects between 1 and MAX_DYNAMIC_OBJECTS
+    int objectsNumber = randomNumber(1, MAX_DYNAMIC_OBJECTS);
+    for(int i = 0 ; i < objectsNumber ; i++)
+    {
```
Comment (mthl) [inline enki/RandomWorld.cpp]: like in the `if` case, a space should be present after the `for` or any control structure (however not function calls)

File: enki/RandomWorld.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#define ONLY_SIMPLE_SHAPES 1
+#define MAX_HEIGHT 800
+#define MIN_HEIGHT 200
+#define MAX_WIDTH 800
+#define MIN_WIDTH 200
+#define MIN_RADIUS 300
+#define MAX_RADIUS 500
+
+#define NUMBER_OF_ROBOTS_TYPES 5
+enum ROBOT_TYPES { THYMIO2, EPUCK, SBOT, MARXBOT, KHEPERA };
```
Comment (mthl) [inline enki/RandomWorld.h]: maybe use `const int`?

File: enki/RandomWorld.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+#include <enki/PhysicalEngine.h>
+#include <enki/robots/thymio2/Thymio2.h>
+#include <enki/robots/e-puck/EPuck.h>
+#include <enki/robots/s-bot/Sbot.h>
+#include <enki/robots/marxbot/Marxbot.h>
+#include <enki/robots/khepera/Khepera.h>
+#include <enki/Random.h>
+
+//TODO: Explain goblals vars
+/** MAX_ROBOTS*/
```
Comment (mthl) [inline enki/RandomWorld.h]: "Explain goblals vars" => "Explain constants meaning"

File: tests/enkiTestRandomWorld.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+{
+    SECTION ( " A random number" )
+    {
+        for ( int i = 0 ; i < ITERATION_NUMBER ; i++ )
+        {
+            int a = randomNumber(0, i);
+            REQUIRE ( (a >= 0 && a <= i) );
+        }
+    }
+    SECTION ( " A random color" )
```
Comment (mthl) [inline tests/enkiTestRandomWorld.cpp]: Can you use the same coding style as used in `tests/enkiTestWorld.cpp`?

File: tests/enkiTestRandomWorld.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
+/* Enki - a fast 2D robot simulator
+   Copyright © 2017 Sebastien Pouteau <sebastien.pouteau@etu.u-bordeaux.fr>
+   Copyright © 2017 Mathieu Lirzin <mathieu.lirzin@etu.u-bordeaux.fr>
+
```
Comment (mthl) [inline tests/enkiTestRandomWorld.cpp]: I don't think me or @sebpouteau was involved with this.  Please add a copyright notice for yourself instead

File: tests/testDisplayRandomWorld.cpp [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#include <viewer/Viewer.h>
+#include <QApplication>
+#include <enki/PhysicalEngine.h>
+#include <enki/RandomWorld.h>
+
+int main(int argc, char *argv[])
```
Comment (mthl) [inline tests/testDisplayRandomWorld.cpp]: Maybe add a short explanation what this file is meant for.  What about something like:

"Manual test for generating a random world and displaying it."

File: tests/testDisplayRandomWorld.cpp [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#include <viewer/Viewer.h>
```
Comment (mthl) [inline tests/testDisplayRandomWorld.cpp]: IMO you can add Copyright notice + copying statement even if the code is trivial.

File: tests/enkiTestRandomWorld.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+// Maybe it makes more sens to let this as a #define ?
+#define DEBUG 0
+
+// Iteration number used in loops to check if the random
+// module is robust.
+const int ITERATION_NUMBER = 100000;
+
+TEST_CASE( "UNIT Testing", "[RandomModule]" )
+{
+	SECTION ( " A random number" )
```
Comment (sebpouteau) [inline tests/enkiTestRandomWorld.cpp]: Extra space : " A random number" -> "A random number"

File: tests/enkiTestRandomWorld.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+	{
+		for ( int i = 0 ; i < ITERATION_NUMBER ; i++ )
+		{
+			int a = randomNumber(0, i);
+			REQUIRE ( (a >= 0 && a <= i) );
+		}
+	}
+	SECTION ( " A random color" )
```
Comment (sebpouteau) [inline tests/enkiTestRandomWorld.cpp]: Extra space : " A random color" -> "A random color"

File: enki/RandomWorld.cpp [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+			r = randomMarxbot(world);
+			break;
+		case KHEPERA:
+			r = randomKhepera(world);
+			break;
+	}
+	return r;
+}
+
+Thymio2* randomThymio(World* world)
```
Comment (sebpouteau) [inline enki/RandomWorld.cpp]: Warning! You don't return a thymio !

Comment (nicpalard) [inline enki/RandomWorld.cpp]: Thanks, it was corrected on the merge you did but I didn't get that version ^^'

File: tests/enkiTestRandomWorld.cpp [hunk 14]
```diff
@@ -38,15 +38,15 @@ const int ITERATION_NUMBER = 100000;
 
 TEST_CASE( "UNIT Testing", "[RandomModule]" )
 {
-	SECTION ( " A random number" )
+	SECTION ( "A random number" )
 	{
 		for ( int i = 0 ; i < ITERATION_NUMBER ; i++ )
```
Comment (mthl) [inline tests/enkiTestRandomWorld.cpp]: As I understand it from [Aseba tests](https://github.com/aseba-community/aseba/blob/master/tests/test-http.cpp),  Catch related syntax should have their `{` bracket on the same line as the `)`.

File: tests/enkiTestRandomWorld.cpp [hunk 15]
```diff
@@ -38,15 +38,15 @@ const int ITERATION_NUMBER = 100000;
...
 		{
 			int a = randomNumber(0, i);
 			REQUIRE ( (a >= 0 && a <= i) );
 		}
 	}
-	SECTION ( " A random color" )
+	SECTION ( "A random color" )
 	{
```
Comment (mthl) [inline tests/enkiTestRandomWorld.cpp]: `SECTION` or any Catch keyword should not have a space after them.

File: enki/RandomWorld.cpp [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+
+// NOTE: UniformRand() from Random.h doesn't produce random values as expected.
+// /!\ This may be slower
+int randomNumber(int min, int max)
+{
+	std::random_device rd; // random number
+	std::mt19937 eng(rd()); // seed
+	std::uniform_int_distribution<> distr(min, max); // range
+	return distr(eng);
+	//return UniformRand(min, max)();
```
Comment (Krazyxx) [inline enki/RandomWorld.cpp]: This comment should be deleted.

Comment (nicpalard) [inline enki/RandomWorld.cpp]: No this shouldn't be deleted because it's the right way to do but it's kinda buggy right now but i'll keep it as a reminder

Comment (mthl) [inline enki/RandomWorld.cpp]: I think it is fine to keep this commented line as long as you provide a rationale as a comment inside the code too.

File: enki/Random.h [hunk 17]
```diff
@@ -7,8 +7,8 @@
     Copyright (C) 2006-2008 Laboratory of Robotics Systems, EPFL, Lausanne
     See AUTHORS for details
 
-    This program is free software; the authors of any publication
-    arising from research using this software are asked to add the
+    This program is free software; the authors of any publication
```
Comment (nicpalard) [inline enki/Random.h]: All this file wasn't really modified, it was just to revert a mistake I did in a previous commit in this PR (c5f2431bb29fa749f0fad70f2dc95e9c1f041759), don't take this file into account.

File: tests/CMakeLists.txt [hunk 18]
```diff
@@ -33,7 +33,6 @@ add_test(enki_test_Marxbot enkiTestWorld "Marxbot")
 add_test(enki_test_Khepera enkiTestWorld "Khepera")
 add_test(enki_test_Sbot enkiTestWorld "S-Bot")
 
-add_executable(enkiTestRandomWorld enkiTestRandomWorld.cpp)
-target_link_libraries(enkiTestRandomWorld catch enki)
-add_test(enki_random enkiTestRandomWorld)
-
+add_executable(randomworld randomWorld.cpp Randomizer.cpp)
+target_link_libraries(randomworld catch enki)
+add_test(random_world enkiTestRandomWorld)
```
Comment (mthl) [inline tests/CMakeLists.txt]: "enkiTestRandomWorld" should be replace with "randomworld", no?

File: enki/CMakeLists.txt [hunk 19]
```diff
@@ -25,10 +24,10 @@ add_library(enki
 	robots/thymio2/Thymio2.cpp
 )
 
-set_target_properties(enki PROPERTIES VERSION ${LIB_VERSION_STRING} 
+set_target_properties(enki PROPERTIES VERSION ${LIB_VERSION_STRING}
                                         SOVERSION ${LIB_VERSION_MAJOR})
```
Comment (mthl) [inline enki/CMakeLists.txt]: This indentation change should be removed

File: tests/randomWorld.cpp [hunk 20]
```diff
@@ -28,15 +28,13 @@
...
 #define DEBUG 0
 
-// Iteration number used in loops to check if the random
-// module is robust.
+// Used to ensure the robustness of the tested code.
 const int ITERATION_NUMBER = 100000;
 
-TEST_CASE( "UNIT Testing", "[RandomModule]" )
+TEST_CASE ( "UNIT Testing", "[Enki::RandomWorld.h]" )
 {
```
Comment (mthl) [inline tests/randomWorld.cpp]: nitpick: Like previously stated, Catch related keyword doesn't have a space before the parenthese, and the `{` should be on the same line of the closing parenthese.

File: tests/randomWorld.cpp [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+#include <iostream>
+
+/*!
+   \file enkiTestRandomWorld.cpp
+   \brief UNIT tests for random module using BDD style
+
+   This file aims to test the random module features.
+*/
+
+
```
Comment (sebpouteau) [inline tests/randomWorld.cpp]: Skip the line in excess, Thanks

File: tests/randomWorld.cpp [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+		{
+			REQUIRE ( (w->r <= MAX_RADIUS && w->r >= MIN_RADIUS) );
+			REQUIRE ( w->objects.size() <= (MAX_DYNAMIC_OBJECTS + MAX_ROBOTS) );
+			if (DEBUG)
+			{
+				std::cout << "[CIRCULAR WORLD]" << std::endl;
+				std::cout << "- radius : " << w->r << std::endl;
+			}
+		}
+
```
Comment (nicpalard) [inline tests/randomWorld.cpp]: Remove extra space

**Comment by mthl** [comment]
applied as e7c0a4b12b124e84c372bda39e2564c3ba0eae86

**Pull request closed without merging**